### ISSUE TITLE
JsonStringEnumMemberNameAttribute does not support empty strings.

### ISSUE THREAD

**Issue by Drake53** [dotnet/runtime#107367]
### Description

For my use case I have the following third party XSD (simplified) for which I generate an enum:
```xml
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:simpleType name="YesOrNoOrEmpty">
        <xs:restriction base="xs:string">
            <xs:enumeration value="y" />
            <xs:enumeration value="n" />
            <xs:enumeration value="" />
        </xs:restriction>
    </xs:simpleType>
</xs:schema>
```

I want the generated enum to be serializable in both XML and JSON, so I annotate the members with both XmlEnumAttribute as the new JsonStringEnumMemberNameAttribute from System.Text.Json v9-preview.

For XML this works fine, but with JSON I get an exception, because empty strings are not allowed. I assume this might have something to do with support for Flags enums? But for enums without FlagsAttribute this restriction should not apply.

### Reproduction Steps

```csharp
using System.Text.Json.Serialization;
using System.Xml.Serialization;

public enum YesOrNoOrEmpty
{
    [JsonStringEnumMemberName("y")]
    [XmlEnum("y")]
    Yes,

    [JsonStringEnumMemberName("n")]
    [XmlEnum("n")]
    No,

    [JsonStringEnumMemberName("")]
    [XmlEnum("")]
    Empty,
}
```

### Expected behavior

`JsonStringEnumConverter` is able to serialize and deserialize enums where the string representation of one of the enum members is an empty string. This matches behaviour of System.Xml's XmlEnumAttribute, and EnumMemberAttribute which is used by Newtonsoft.Json.

### Actual behavior

`JsonStringEnumConverter` (using `EnumConverter` internally) does a `string.IsNullOrEmpty` check and throws an InvalidOperationException: `Enum type '{0}' uses unsupported identifier '{1}'. It must not be null, empty, or containing leading or trailing whitespace. Flags enums must additionally not contain commas.`

### Known Workarounds

Implementing a custom JsonConverter for enums

**Comment by eiriktsarpalis** [comment]
There's nothing technically preventing the use of empty strings (even in the case of flags, where you might be able to write something like `y, , n`). [It got added intentionally](https://github.com/dotnet/runtime/blob/49bf719618acd3444c0b6b58bc9f7661a8c25708/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/EnumConverterTests.cs#L1174-L1209) intentionally principally to avoid pathological outputs (like the one mentioned above) and discourage it from being used. 

We might consider relaxing that restriction in the future, provided there is sufficient demand for it.

**Comment by Drake53** [comment]
> There's nothing technically preventing the use of empty strings (even in the case of flags, where you might be able to write something like `y, , n`). [It got added intentionally](https://github.com/dotnet/runtime/blob/49bf719618acd3444c0b6b58bc9f7661a8c25708/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/EnumConverterTests.cs#L1174-L1209) principally to avoid pathological outputs (like the one mentioned above) and discourage it from being used.
> 
> We might consider relaxing that restriction in the future, provided there is sufficient demand for it.

> There's nothing technically preventing the use of empty strings (even in the case of flags, where you might be able to write something like `y, , n`).

It's true the example you give could work, even though it looks a bit weird, but the problem is that the empty string in json will become ambiguous, because it can be either no flags, or only the empty string member's flag:
```csharp
[Flags]
public enum FlagsEnum
{
    [JsonStringEnumMemberName("")]
    Empty,
    [JsonStringEnumMemberName("1")]
    One,
}

JsonSerializer.Serialize(FlagsEnum.Empty); // returns ""
JsonSerializer.Serialize((FlagsEnum)0); // returns ""

JsonSerializer.Deserialize<FlagsEnum>(""); // ambiguous
```

**Comment by IvanJosipovic** [comment]
I would also like the library to support [JsonStringEnumMemberName("")]

**Comment by josephcrivello** [comment]
I also would like `JsonStringEnumMemberName` to support empty and null strings.

> There's nothing technically preventing the use of empty strings (even in the case of flags, where you might be able to write something like `y, , n`). [It got added intentionally](https://github.com/dotnet/runtime/blob/49bf719618acd3444c0b6b58bc9f7661a8c25708/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/EnumConverterTests.cs#L1174-L1209) principally to avoid pathological outputs (like the one mentioned above) and discourage it from being used.
> 
> We might consider relaxing that restriction in the future, provided there is sufficient demand for it.

This feels like a **bug** to me, not an **enhancement request**.

In a perfect world, I understand the desire to discourage cases like the one you outlined. But one of the most common use cases for `System.Text.Json.Serialization` is writing .NET applications and libraries that interact with restful APIs. Typically, the developer or organization responsible for this task has no control over the JSON schema defined by the API, or it cannot be easily changed.

Isn't `System.Text.Json.Serialization` supposed to be able to support the serialization and deserialization of any well-formed JSON?

If yes, then it seems like this behavior is incorrect. `JsonStringEnumMemberName` is imposing an opinionated limitation on serializing and deserializing well-formed JSON with no technical basis or rationale.

**Comment by eiriktsarpalis** [comment]
> Isn't System.Text.Json.Serialization supposed to be able to support the serialization and deserialization of any well-formed JSON?

Any serializer that extracts a JSON contract from types will by definition do so in an opinionated manner that doesn't accommodate every conceivable use case. For any such corner cases the recommended workaround is to employ a custom converter, if and until a particular corner case becomes supported in the built-in converters.

**Comment by josephcrivello** [comment]
> Any serializer that extracts a JSON contract from types will by definition do so in an opinionated manner that doesn't accommodate every conceivable use case. For any such corner cases the recommended workaround is to employ a custom converter, if and until a particular corner case becomes supported in the built-in converters.

Of course, all serializers should make their own opinionated design decisions. This is useful because it helps ensure a cohesive library design.

I am under the impression from the documentation that the following are the key "opinions" underlying the design of `System.Text.Json`:

> The System.Text.Json library design emphasizes high performance and low memory allocation over an extensive feature set. Built-in UTF-8 support optimizes the process of reading and writing JSON text encoded as UTF-8, which is the most prevalent encoding for data on the web and files on disk.
> [JSON serialization and deserialization (marshalling and unmarshalling) in .NET - overview](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/overview)

I am aware that there are other goals that are not mentioned in this brief description, such as achieving AOT support.

Is preventing the consumer of the library from producing _pathological outputs_ one of the opinionated design goals for the `System.Text.Json` library?

**Comment by eiriktsarpalis** [comment]
> Is preventing the consumer of the library from producing "pathological outputs" one of the opinionated design goals for the System.Text.Json library?

Yes. There are a number of things that the library explicitly does not permit _out of the box_ in the interest of performance, security, promoting good practices, and adhering to the principle of least surprise. For example, it doesn't allow unconstrained polymorphism nor does it do implicit type coercion. The reason why whitespace isn't allowed in enum identifiers is because of the ambiguity that it introduces in the context of flag enum serialization.

That does not mean they're not possible to achieve using customization, and we make sure that the library exposes all the necessary extensibility points in the likely event that the built-in features don't support a specific scenario.

**Comment by josephcrivello** [comment]
> Yes. There are a number of things that the library explicitly does not permit _out of the box_ in the interest of performance, security, promoting good practices, and adhering to the principle of least surprise. For example, it doesn't allow unconstrained polymorphism nor does it do implicit type coercion. The reason why whitespace isn't allowed in enum identifiers is because of the ambiguity that it introduces in the context of flag enum serialization.

Ok, all that seems pretty reasonable. Thank you for taking the time to respond, your response has helped me clarify the nature of my disagreement. Here is my precise objection:

An empty string is _not_ a pathological output, other than in the noted special case of flags enum serialization.

An empty string is a common, non-pathological use case for indicating the absence of a meaningful value. In this use case, makes perfect sense to map the default value of an enum to an empty string.

It seems like the baby is being thrown out with the bathwater.

I understand that no serialization library can address every use case, but it is a pity because so many restful APIs use empty strings in this manner. Integrating .NET applications and libraries with restful APIs is presumably one of (or the) most common use cases for the serializer. Now, .NET 9 is about to be released, and the die will be cast on the expected behavior for `JsonStringEnumMemberNameAttribute` presumably making it difficult or impossible to change in the future.

I'll stop beating a dead horse now.

**Comment by eiriktsarpalis** [comment]
> An empty string is not a pathological output, other than in the noted special case of flags enum serialization.

I don't think we're in disagreement here. As mentioned earlier, it is conceivable that the converter could be made to be more forgiving provided that flags aren't being used. However, this isn't something we can change with .NET 9 for which development concluded back in August.

**Comment by IvanJosipovic** [comment]
> > We might consider relaxing that restriction in the future, provided there is sufficient demand for it.
> 
> Since you said you might reconsider if there is enough demand, I'll give some feedback on that: I ran into this exact issue earlier today.
> 
> I have a third-party defined json schema (I can't influence them) using an enum with the empty string to represent "none". There are no flags involved, because this scenario is single choice.
> 
> There is a work-around, so I can make the code work. But I expected .NET to support this :(

@yieldOnly, do you mind posting the work-around?

**Comment by 0xced** [comment]
The [Docker Engine API](https://docs.docker.com/reference/api/engine/) has many empty strings which naturally map onto enums.

For example, the [SystemInfo Isolation](https://docs.docker.com/reference/api/engine/version/v1.50/#tag/System/operation/SystemInfo)  is documented as Enum: `"default"` `"hyperv"` `"process"` `""` where the empty string is listed as a legitimate value.

> We might consider relaxing that restriction in the future, provided there is sufficient demand for it.

Let's hope that my 11th 👍 will help reach the demand threshold. 🤞

**Comment by ElDuderinoBerlin** [comment]
Please implement this feature.

**Comment by DavidM29** [comment]
To add one more reason, Microsoft's own product, `Business Central`, uses `" "` as a default value in enums, so it would be nice to have this without workarounds.

**Comment by tomupson** [comment]
+1 from me - facing this in similar scenarios to others where "none" / "unknown" values are represented as empty strings
